### ISSUE TITLE
Unexpected matching logic

### ISSUE THREAD

**Issue by mantoni** [sinonjs/referee#54]
To my surprise, the following passes:

```js
assert.match(
  { array: [1, 2, 3], more: true },
  { array: [2] }
)
```

My expectation would be that the above would fail due to the array not being equal. The following should pass:

```js
assert.match(
  { array: [1, 2, 3], more: true },
  { array: [1, 2, 3] }
)
```

This differs from how Sinon.JS matchers work:

```js
const spy = sinon.spy()
spy({ array: [1, 2, 3], more: true })

spy.calledWithMatch({ array: [1, 2, 3] }) // true
spy.calledWithMatch({ array: [2] }) // false
```

I prefer the Sinon match logic, because it allows me to explicitly state what I expect with nested matchers:

```js
spy.calledWithMatch({ array: sinon.match.array.contains([2]) })
```

If we're [merging the matcher logic](#29), I think we should basically replace the referee logic with the Sinon implementation, especially since the user base is a tad larger over there.

**Comment by mroderick** [comment]
Good point. We should probably look into finishing that work

**Comment by mroderick** [comment]
I'll see if I can get it closer this weekend, as I am having a hack day with @khamiltonuk

**Comment by mantoni** [comment]
Oh nice! That'd be awesome. Unfortunately, I don't have the time to look into the implementation myself rn, but I'm happy to review and assist.

**Comment by mroderick** [comment]
Is this still broken?

**Comment by mantoni** [comment]
Yes, this is also relevant with the `next` branch of samsam. The matching logic didn't change.

**Comment by fatso83** [comment]
Seems this failed to update the docs, so the docs are broken. Ref https://github.com/sinonjs/sinon/discussions/2614#discussioncomment-11158654
